unsloth/utils/attention_dispatch.py

```diff
@@ -0,0 +1,271 @@
+        invalid from one forward pass to the next (e.g., switching from training to
+        inference). For now, I'm keeping separate from AttentionContext for the sake of
+        better grouping of params.
+    """
+
+    backend: str
+    n_kv_heads: int
+    n_groups: int
+    flash_dense_kwargs: Optional[dict[str, Any]] = None
+    flash_varlen_kwargs: Optional[dict[str, Any]] = None
+    sdpa_kwargs: Optional[dict[str, Any]] = None
+    xformers_kwargs: Optional[dict[str, Any]] = None
+
+
+@dataclass
+class AttentionContext:
+    """Per-call info required to run attention."""
+
+    bsz: int
+    q_len: int
+    kv_seq_len: int
+    n_heads: int
+    head_dim: int
+    requires_grad: bool
+    seq_info: Optional[Tuple[Tensor, Tensor, int]]
+    attention_mask: Optional[Tensor]
+    causal_mask: Optional[Any]
+    sliding_window: Optional[int] = None
+
+
+def select_attention_backend(use_varlen: bool = False) -> str:
+    """Return attention backend based on availability / priority order."""
+
+    if HAS_FLASH_ATTENTION:
+        if use_varlen:
+            return FLASH_VARLEN
+        else:
+            return FLASH_DENSE
+    if HAS_XFORMERS:
+        return XFORMERS
+    return SDPA
+
+
+def run_attention(
+    *,
+    config: AttentionConfig,
+    context: AttentionContext,
+    Q: Tensor,
+    K: Tensor,
+    V: Tensor,
+) -> Tensor:
+    """Run attention using config / context info."""
+
+    backend = config.backend
+    if backend == FLASH_VARLEN and context.seq_info is None:
+        backend = FLASH_DENSE if HAS_FLASH_ATTENTION else SDPA
+    flash_dense_kwargs = config.flash_dense_kwargs or {}
+    flash_varlen_kwargs = config.flash_varlen_kwargs or {}
+    sdpa_kwargs = config.sdpa_kwargs or {}
+    xformers_kwargs = config.xformers_kwargs or {}
+
+    bsz = context.bsz
+    n_heads = context.n_heads
+    q_len = context.q_len
+    head_dim = context.head_dim
+    kv_seq_len = context.kv_seq_len
+    requires_grad = context.requires_grad
+    sliding_window = context.sliding_window
+
+    if backend == FLASH_VARLEN:
+        Q_f = Q.transpose(1, 2).reshape(bsz * q_len, n_heads, head_dim)
+        K_f = K.transpose(1, 2).reshape(bsz * q_len, config.n_kv_heads, head_dim)
+        V_f = V.transpose(1, 2).reshape(bsz * q_len, config.n_kv_heads, head_dim)
+        _, cu_seqlens, max_seqlen = context.seq_info
+        return flash_attn_varlen_func(
+            Q_f,
+            K_f,
+            V_f,
+            cu_seqlens,
+            cu_seqlens,
+            max_seqlen,
+            max_seqlen,
+            **flash_varlen_kwargs,
+        ).view(bsz, q_len, n_heads, head_dim)
+    elif backend == FLASH_DENSE:
+        Q_t = Q.transpose(1, 2)
+        K_t = K.transpose(1, 2)
+        V_t = V.transpose(1, 2)
+        return flash_attn_func(Q_t, K_t, V_t, **flash_dense_kwargs).reshape(
+            bsz, q_len, n_heads, head_dim
+        )
+    elif backend == XFORMERS:
+        attn_bias = build_xformers_block_causal_mask(
```

Comment:
@djsaunde Oh I meant here for caching - isn't this called like say 24 times if there's 24 layers? Shouldn't we cache this?